File: cpp/DPR.cpp
#include <iostream>
#include <iomanip>
#include "DPR.hpp"

using namespace std;

DPR::DPR() {
    this->maksLebarId = 3;
    this->maksLebarNama = 5;
    this->maksLebarBidang = 7;
    this->maksLebarPartai = 7;
}

DPR::~DPR() {
}

void DPR::setMaksLebarId(unsigned int maksLebarId) {
    this->maksLebarId = maksLebarId;
}

void DPR::setMaksLebarNama(unsigned int maksLebarNama) {
    this->maksLebarNama = maksLebarNama;
}

void DPR::setMaksLebarBidang(unsigned int maksLebarBidang) {
    this->maksLebarBidang = maksLebarBidang;
}

void DPR::setMaksLebarPartai(unsigned int maksLebarPartai) {
    this->maksLebarPartai = maksLebarPartai;
}

void DPR::setDaftarAnggotaDPR(vector<AnggotaDPR> daftarAnggotaDPR) {
    this->daftarAnggotaDPR = daftarAnggotaDPR;
}

unsigned int DPR::ambilMaksLebarId() {
    return this->maksLebarId;
}

unsigned int DPR::ambilMaksLebarNama() {
    return this->maksLebarNama;
}

unsigned int DPR::ambilMaksLebarBidang() {
    return this->maksLebarBidang;
}

unsigned int DPR::ambilMaksLebarPartai() {
    return this->maksLebarPartai;
}

vector<AnggotaDPR> DPR::ambilDaftarAnggotaDPR() {
    return daftarAnggotaDPR;
}

void DPR::konfigurasiTabel(AnggotaDPR anggotaBaru) {
    setMaksLebarNama(maksimum(this->ambilMaksLebarNama(), anggotaBaru.ambilNama().size() + 1));
    setMaksLebarBidang(maksimum(this->ambilMaksLebarBidang(), anggotaBaru.ambilBidang().size() + 1));
    setMaksLebarPartai(maksimum(this->ambilMaksLebarPartai(), anggotaBaru.ambilPartai().size() + 1));
}

void DPR::tambahAnggota(AnggotaDPR anggotaBaru) {
    // pastikan tidak ada id yang sama
    unsigned int indexAnggotaDPRDicari = cariAnggota(anggotaBaru.ambilId());
    if(indexAnggotaDPRDicari == daftarAnggotaDPR.size()) {
        this->daftarAnggotaDPR.push_back(anggotaBaru);

        setMaksLebarId(maksimum(this->ambilMaksLebarId(), hitungPanjangUnsignedInt(anggotaBaru.ambilId()) + 1));
        konfigurasiTabel(anggotaBaru);
    } else {
        cout << "ditemukan id sama\n";
    }
}

unsigned int DPR::maksimum(unsigned int a, unsigned int b) {
    return (a > b ? a : b);
}

unsigned int DPR::hitungPanjangUnsignedInt(unsigned int bil) {
    unsigned int totalDigit = 0;
    if(bil == 0) totalDigit = 1;
    else {
        while(bil != 0) {
            bil /= 10;
            totalDigit++;
        }
    }

    return totalDigit;
};

unsigned int DPR::cariAnggota(unsigned int idAnggota) {
    // cari terlebih dahulu anggota berdasarkan id
    unsigned int idDitunjuk;
    int indeks = 0, banyakAnggota = daftarAnggotaDPR.size();
    bool ditemukan = 0;

    while(indeks < banyakAnggota && !ditemukan) {
        if(daftarAnggotaDPR[indeks].ambilId() == idAnggota) {
            ditemukan = 1;
        } else {
            indeks++;
        }
    }

    return indeks;
}

void DPR::ubahDataAnggota(unsigned int idAnggota, AnggotaDPR dataAnggota) {
    // cari terlebih dahulu anggota berdasarkan id
    unsigned int indexAnggotaDPRDicari = cariAnggota(idAnggota);

    // jika ditemukan, perbarui data anggota
    if(indexAnggotaDPRDicari != daftarAnggotaDPR.size()) {
        daftarAnggotaDPR[indexAnggotaDPRDicari].setNama(dataAnggota.ambilNama());
        daftarAnggotaDPR[indexAnggotaDPRDicari].setBidang(dataAnggota.ambilBidang());
        daftarAnggotaDPR[indexAnggotaDPRDicari].setPartai(dataAnggota.ambilPartai());

        konfigurasiTabel(dataAnggota);
    } else {
        cout << "tidak ditemukan id sama\n";
    }
}

void DPR::hapusDataAnggota(unsigned int idAnggota) {
    // cari terlebih dahulu anggota berdasarkan id
    unsigned int indexAnggotaDPRDicari = cariAnggota(idAnggota);

    // jika ditemukan, hapus data anggota
    if(indexAnggotaDPRDicari != daftarAnggotaDPR.size()) {
        daftarAnggotaDPR.erase(daftarAnggotaDPR.begin() + indexAnggotaDPRDicari);

        // perbarui maksimum dengan mengecek keseluruhan
        this->maksLebarId = 3;
        this->maksLebarNama = 5;
        this->maksLebarBidang = 7;
        this->maksLebarPartai = 7;

        for(AnggotaDPR anggota: daftarAnggotaDPR) {
            setMaksLebarId(maksimum(this->ambilMaksLebarId(), hitungPanjangUnsignedInt(anggota.ambilId()) + 1));
            konfigurasiTabel(anggota);
        }
    } else {
        cout << "tidak ditemukan id sama\n";
    }
}

void DPR::printGarisPendek(unsigned int panjang, bool prefix, bool postfix) {
    if(prefix) cout << "+";
    for(unsigned int i = 0; i < panjang; i++) cout << "-";
    if(postfix) cout << "+\n";
}

void DPR::printGaris() {
    printGarisPendek(maksLebarId + 1, 1, 0);
    printGarisPendek(maksLebarNama + 1, 1, 0);
    printGarisPendek(maksLebarBidang + 1, 1, 0);
    printGarisPendek(maksLebarPartai + 1, 1, 1);
}

void DPR::tampilkan() {
    // tampilkan header
    printGaris();
    cout << left << setw(maksLebarId) << "| id "
                    << setw(maksLebarNama) << "| nama "
                    << setw(maksLebarBidang) << "| bidang "
                    << setw(maksLebarPartai) << "| partai |\n";
    printGaris();

    // tampilkan data
    for(AnggotaDPR anggota: daftarAnggotaDPR) {
        cout << left << "| " << setw(maksLebarId) << anggota.ambilId()
                            << "| " << setw(maksLebarNama) << anggota.ambilNama()
                            << "| " << setw(maksLebarBidang) << anggota.ambilBidang()
                            << "| " << setw(maksLebarPartai) << anggota.ambilPartai() << "|\n";
    }

    if(!daftarAnggotaDPR.empty()) printGaris();
    cout << endl;
}

File: python/DPR.py
from AnggotaDPR import AnggotaDPR

class DPR:
    def __init__(self):
        self.maksLebarId = 3
        self.maksLebarNama = 5
        self.maksLebarBidang = 7
        self.maksLebarPartai = 7
        self.daftarAnggotaDPR = []

    def setMaksLebarId(self, maksLebarId):
        self.maksLebarId = maksLebarId

    def setMaksLebarNama(self, maksLebarNama):
        self.maksLebarNama = maksLebarNama

    def setMaksLebarBidang(self, maksLebarBidang):
        self.maksLebarBidang = maksLebarBidang

    def setMaksLebarPartai(self, maksLebarPartai):
        self.maksLebarPartai = maksLebarPartai

    def setDaftarAnggotaDPR(self, daftarAnggotaDPR):
        self.daftarAnggotaDPR = daftarAnggotaDPR

    def ambilMaksLebarId(self):
        return self.maksLebarId

    def ambilMaksLebarNama(self):
        return self.maksLebarNama

    def ambilMaksLebarBidang(self):
        return self.maksLebarBidang

    def ambilMaksLebarPartai(self):
        return self.maksLebarPartai

    def ambilDaftarAnggotaDPR(self):
        return self.daftarAnggotaDPR

    def konfigurasiTabel(self, anggotaBaru: AnggotaDPR):
        self.setMaksLebarNama(max(self.ambilMaksLebarNama(), len(anggotaBaru.ambilNama()) + 1))
        self.setMaksLebarBidang(max(self.ambilMaksLebarBidang(), len(anggotaBaru.ambilBidang()) + 1))
        self.setMaksLebarPartai(max(self.ambilMaksLebarPartai(), len(anggotaBaru.ambilPartai()) + 1))

    def tambahAnggota(self, anggotaBaru: AnggotaDPR):
        # pastikan tidak ada id yang sama
        if anggotaBaru.ambilId() not in [anggota.ambilId() for anggota in self.daftarAnggotaDPR]:
            self.daftarAnggotaDPR.append(anggotaBaru)
            self.setMaksLebarId(max(self.ambilMaksLebarId(), len(str(anggotaBaru.ambilId())) + 1))
            self.konfigurasiTabel(anggotaBaru)
        else:
            print("ditemukan id sama")

    def cariAnggota(self, idAnggota: int):
        index = 0
        for anggota in self.daftarAnggotaDPR:
            if anggota.ambilId() == idAnggota:
                return index
            index += 1
        return len(self.daftarAnggotaDPR)

    def ubahDataAnggota(self, idAnggota: int, dataAnggota: AnggotaDPR):
        indexAnggotaDPRDicari = self.cariAnggota(idAnggota)
        if indexAnggotaDPRDicari != len(self.daftarAnggotaDPR):
            self.daftarAnggotaDPR[indexAnggotaDPRDicari].setNama(dataAnggota.ambilNama())
            self.daftarAnggotaDPR[indexAnggotaDPRDicari].setBidang(dataAnggota.ambilBidang())
            self.daftarAnggotaDPR[indexAnggotaDPRDicari].setPartai(dataAnggota.ambilPartai())
            self.konfigurasiTabel(dataAnggota)
        else:
            print("tidak ditemukan id sama")

    def hapusDataAnggota(self, idAnggota: int):
        indexAnggotaDPRDicari = self.cariAnggota(idAnggota)
        if indexAnggotaDPRDicari != len(self.daftarAnggotaDPR):
            del self.daftarAnggotaDPR[indexAnggotaDPRDicari]
            self.maksLebarId = 3
            self.maksLebarNama = 5
            self.maksLebarBidang = 7
            self.maksLebarPartai = 7
            for anggota in self.daftarAnggotaDPR:
                self.setMaksLebarId(max(self.ambilMaksLebarId(), len(str(anggota.ambilId())) + 1))
                self.konfigurasiTabel(anggota)
        else:
            print("tidak ditemukan id sama")

    def tampilkan(self):
        print("{:<{}} {:<{}} {:<{}} {:<{}}".format("id", self.maksLebarId, "nama", self.maksLebarNama,
                                                    "bidang", self.maksLebarBidang, "partai", self.maksLebarPartai))
        for anggota in self.daftarAnggotaDPR:
            print("{:<{}} {:<{}} {:<{}} {:<{}}".format(anggota.ambilId(), self.maksLebarId, anggota.ambilNama(),
                                                        self.maksLebarNama, anggota.ambilBidang(), self.maksLebarBidang,
                                                        anggota.ambilPartai(), self.maksLebarPartai))
        print()
